hazelcast/src/main/java/com/hazelcast/jet/impl/MasterJobContext.java

```diff
@@ -595,10 +595,15 @@ private void onStartExecutionComplete(Throwable error, Collection<Entry<MemberIn
                     // have to use Async version, the future is completed inside a synchronized block
                     .whenCompleteAsync(withTryCatch(logger, (r, e) -> finalizeJob(finalError)));
         } else {
-            if (error instanceof ExecutionNotFoundException) {
-                // If the StartExecutionOperation didn't find the execution, it means that we must have cancelled it.
+            if (error instanceof ExecutionNotFoundException && requestedTerminationMode != null) {
+                // If the StartExecutionOperation didn't find the execution and requestedTerminationMode
+                // is not null it means that we must have cancelled it.
                 // Let's pretend that the StartExecutionOperation returned JobTerminateRequestedException
-                assert requestedTerminationMode != null && !requestedTerminationMode.isWithTerminalSnapshot();
+                // Note that the local execution cancellation need not be performed only on
+                // an explicit user cancellation request. In this case, requestedTerminationMode
+                // becomes null. For example, this local cancellation can happen in
+                // JobExecutionService#onMemberRemoved. We will continue to handle these cases as if we
+                // would handle an unspecific exception
```

Comment:
The comment contradicts itself: "it means that we must have cancelled it" and then "it might not be only us cancelling it, but also `onMemberRemoved`".

```java
                // If the StartExecutionOperation didn't find the execution, it means that it was cancelled.
                // This can be either because the master cancelled it, or it can also happen that some participant
                // left and the target cancelled the execution locally in JobExecutionService.onMemberRemoved().
                // Let's pretend that the StartExecutionOperation returned JobTerminateRequestedException.
```

Reply:
I split the comment block and updated the comment as you wanted. Thanks for the suggestion.